admin/google_search_console/class-gwt-client.php

```diff
@@ -0,0 +1,254 @@
+	 * @param int    $accepted_response_code
+	 *
+	 * @return mixed
+	 */
+	public function decode_response( $response, $accepted_response_code = 200 ) {
+		if ( $accepted_response_code === $response->getResponseHttpCode() ) {
+			return json_decode( $response->getResponseBody() );
+		}
+	}
+
+	/**
+	 * Getting the response code, saved from latest request to Google
+	 * @return mixed
+	 */
+	public function get_http_response_code() {
+		return $this->http_response_code;
+	}
+
+	/**
+	 * Clears the options and revokes the token
+	 */
+	public function clear_data() {
+		$this->revokeToken();
+
+		delete_option( self::OPTION_ACCESS_TOKEN );
+		delete_option( self::OPTION_REFRESH_TOKEN );
+	}
+
+	/**
+	 * Initialize the config, will merge given config with default config to be sure all settings are available
+	 *
+	 * @param array $config
+	 */
+	protected function init_config( array $config ) {
+		$config = array_merge( $config, $this->default_config );
+
+		if ( ! empty( $config['application_name'] ) ) {
+			$this->setApplicationName( $config['application_name'] );
+		}
+
+		if ( ! empty( $config['client_id'] ) ) {
+			$this->setClientId( $config['client_id'] );
+		}
+
+		if ( ! empty( $config['client_secret'] ) ) {
+			$this->setClientSecret( $config['client_secret'] );
+		}
+
+		// Set our settings.
+		$this->setRedirectUri( $config['redirect_uri'] );
+		$this->setScopes( $config['scopes'] );
+		$this->setAccessType( 'offline' );
+	}
+	/**
+	 * Refreshing the tokens
+	 */
+	protected function refresh_tokens() {
+		if ( ( $refresh_token = $this->get_refresh_token() ) !== '' && $this->access_token_expired() ) {
+			try {
+				// Refresh the token.
+				$this->refreshToken( $refresh_token );
+
+				$response = $this->getAuth()->token;
+
+				// Check response and if there is an access_token.
+				if ( ! empty( $response ) && ! empty ( $response['access_token'] ) ) {
+					$this->save_access_token( $response );
+				}
+			}
+			catch ( Exception $e ) {
+				return false;
+			}
+		}
+	}
+
+	/**
+	 * Save the refresh token
+	 *
+	 * @param string $refresh_token
+	 */
+	protected function save_refresh_token( $refresh_token ) {
+		update_option( self::OPTION_REFRESH_TOKEN, trim( $refresh_token ) );
+	}
+
+	/**
+	 * Return refresh token
+	 *
+	 * @return string
+	 */
+	protected function get_refresh_token() {
+		return get_option( self::OPTION_REFRESH_TOKEN, '' );
+	}
+
+	/**
+	 * Saving the access token as an option for further use till it expires.
+	 *
+	 * @param array $response
+	 */
+	protected function save_access_token( $response ) {
+		update_option(
+			self::OPTION_ACCESS_TOKEN,
+			array(
+				'refresh_token' => $this->get_refresh_token(),
+				'access_token'  => $response['access_token'],
+				'expires'       => current_time( 'timestamp' ) + $response['expires_in'],
+				'expires_in'    => $response['expires_in'],
+				'created'       => $response['created'],
+			)
+		);
+
+		$this->setAccessToken( json_encode( $response ) );
```

Comment:
We should use `wp_json_encode` here, see https://github.com/Yoast/wordpress-seo/blob/trunk/admin/class-pointers.php#L99-106 for details on how we implemented it before. Because it is only supported in WP 4.1